src/Minsk/CodeAnalysis/Lowering/Lowerer.cs

```diff
@@ -263,12 +222,145 @@ protected override BoundStatement RewriteConditionalGotoStatement(BoundCondition
+        {
+            return new BoundLabelStatement(label);
+        }
+
+        private BoundLiteralExpression Literal(object literal)
+        {
+            Debug.Assert(literal is string || literal is bool || literal is int);
+
+            return new BoundLiteralExpression(literal);
+        }
+
+        private BoundBlockStatement Block(params BoundStatement[] stmts)
+        {
+            return new BoundBlockStatement(ImmutableArray.Create(stmts));
+        }
+
+        private BoundGotoStatement Goto(BoundLabelStatement label)
+        {
+            return new BoundGotoStatement(label.Label);
+        }
+
+        private BoundGotoStatement Goto(BoundLabel label)
+        {
+            return new BoundGotoStatement(label);
+        }
+
+        private BoundConditionalGotoStatement Goto(BoundLabelStatement label, BoundExpression condition, bool jumpIfTrue)
+        {
+            return new BoundConditionalGotoStatement(label.Label, condition, jumpIfTrue);
+        }
+
+        private BoundConditionalGotoStatement GotoTrue(BoundLabelStatement label, BoundExpression condition)
+            => Goto(label, condition, jumpIfTrue: true);
+
+        private BoundConditionalGotoStatement GotoFalse(BoundLabelStatement label, BoundExpression condition)
+            => Goto(label, condition, jumpIfTrue: false);
+
+        private BoundVariableDeclaration VariableDeclaration(VariableSymbol symbol, BoundExpression initExpr)
+        {
+            return new BoundVariableDeclaration(symbol, initExpr);
+        }
+
+        private BoundVariableExpression Variable(VariableSymbol symbol)
+        {
+            return new BoundVariableExpression(symbol);
+        }
+
+        private BoundVariableExpression Variable(BoundVariableDeclaration varDecl)
+        {
+            return new BoundVariableExpression(varDecl.Variable);
+        }
+
+        private BoundVariableDeclaration ConstantDeclaration(string name, BoundExpression initExpr, TypeSymbol? type = null)
+            => VariableDeclarationInternal(name, initExpr, type, isReadOnly: true);
+
+        private BoundVariableDeclaration VariableDeclaration(string name, BoundExpression initExpr, TypeSymbol? type = null)
+            => VariableDeclarationInternal(name, initExpr, type, isReadOnly: false);
+
+        private BoundVariableDeclaration VariableDeclarationInternal(string name, BoundExpression initExpr, TypeSymbol? type, bool isReadOnly)
+        {
+            var symbol = Symbol(name, type ?? initExpr.Type, isReadOnly, initExpr.ConstantValue);
+            return new BoundVariableDeclaration(symbol, initExpr);
+        }
+
+        private LocalVariableSymbol Symbol(string name, TypeSymbol type, bool isReadOnly = true, BoundConstant? constant = null)
+        {
+            return new LocalVariableSymbol(name, isReadOnly, type, constant);
+        }
+
+        private BoundBinaryExpression BinaryExpr(BoundExpression lhs, SyntaxKind kind, BoundExpression rhs)
```

Comment:
I'd name this either `Binary` or `BinaryExpression`. Also, instead of `lhs`, `rhs` we should match the parameter names which are `left` and `right`